cpg_infra/abstraction/google_group_membership.py

```diff
@@ -0,0 +1,207 @@
+
+    member_key: str
+    "The email of the group member"
+
+    member_name: str
+    """
+    a path in the format groups/<group_id>/memberships/<member_id
+    where member_id is the gcloud id of the member, not the email.
+    This is needed when deleting memberships
+    """
+
+
+class GoogleGroupMembershipInputs(object):
+    group_id: Input[str]
+    member_key: Input[str]
+
+    def __init__(self, group_id: str, member_key: str):
+        self.group_id = group_id
+        self.member_key = member_key
+
+
+class GoogleGroupMembershipProviderInputs(TypedDict):
+    group_id: str
+    member_key: str
+
+
+class GoogleGroupMembership(pulumi.dynamic.Resource):
+    """A Pulumi dynamic resource for Google Group memberships."""
+
+    def __init__(
+        self,
+        name: str,
+        props: GoogleGroupMembershipInputs,
+        opts: Optional[ResourceOptions] = None,
+    ) -> None:
+        super().__init__(
+            GoogleGroupMembershipProvider(),
+            name,
+            {**vars(props)},
+            opts,
+        )
+
+
+class GoogleGroupMembershipProvider(pulumi.dynamic.ResourceProvider):
+    """A Pulumi dynamic resource provider for Google Groups settings."""
+
+    def create(self, props: GoogleGroupMembershipProviderInputs):
+        group_id = props['group_id']
+        member_key = props['member_key']
+
+        member = find_member_in_group_by_key(group_id, member_key)
+
+        # If the group membership already exists, then don't try and create it
+        if member != None:
+            return pulumi.dynamic.CreateResult(id_=member['member_name'])
+
+        # Otherwise create it here
+        created_member = add_member_to_group(group_id, member_key)
+
+        return pulumi.dynamic.CreateResult(id_=created_member['member_name'])
+
+    def delete(self, id: str, props: Any):
+        name_parts = id.split('/')
+        group_id = name_parts[1]
+        member = find_member_in_group_by_name(group_id, id)
+
+        # If the member has already been removed then no need to remove
+        if member == None:
+            return
+
+        remove_member_from_group(id)
+
+    def diff(self, id: str, olds: Any, news: Any):
+        # There isn't anything to change on a group membership, they are either a member
+        # or not, so this can always return false
+        return pulumi.dynamic.DiffResult(changes=False)
+
+
+@cache
+def get_groups_service():
+    service: CloudIdentityResource = googleapiclient.discovery.build(  # type: ignore
+        'cloudidentity', 'v1'
+    )
+    return service
+
+
+def find_member_in_group_by_key(group_id: str, member_key: str):
+    member_list = list_group_members(group_id)
+    return next((m for m in member_list if m['member_key'] == member_key), None)
+
+
+def find_member_in_group_by_name(group_id: str, member_name: str):
+    member_list = list_group_members(group_id)
+    return next((m for m in member_list if m['member_name'] == member_name), None)
```

Comment:
I don't _love_ this find by iteration. If it's possible to return a dict from above, and check presence of the member_key in all situations that's my preferred. It's not a big deal, but this would could get called like 10-20k times on a refresh, and some of the group sizes can get quite large.